Earthfile

```
@@ -0,0 +1,417 @@
+
+generate-entrypoints:
+    FROM +builder
+    COPY ./scripts/generate-entrypoints.rb ./generate-entrypoints.rb
+    COPY ./src/checker.mli checker.mli
+    RUN ./generate-entrypoints.rb checker.mli > checkerEntrypoints.ml
+    SAVE ARTIFACT checkerEntrypoints.ml AS LOCAL src/checkerEntrypoints.ml
+    SAVE ARTIFACT checkerEntrypoints.ml /
+    # Image for inline caching
+    SAVE IMAGE --push ghcr.io/tezos-checker/checker/earthly-cache:generate-entrypoints
+
+build-ocaml:
+    FROM +deps-ocaml
+    COPY src/*.ml src/*.mli src/dune ./src/
+    COPY +generate-entrypoints/checkerEntrypoints.ml ./src/
+    COPY tests/*.ml tests/dune ./tests/
+    COPY dune-project ./
+    RUN opam exec -- dune build @install
+    # Image for inline caching
+    SAVE IMAGE --push ghcr.io/tezos-checker/checker/earthly-cache:build-ocaml
+
+build-ligo:
+    FROM +builder
+
+    COPY +ligo-binary/ligo /bin/ligo
+
+    WORKDIR /root
+
+    COPY ./src/*.ml ./src/*.mligo ./src/
+    COPY +generate-entrypoints/checkerEntrypoints.ml ./src/checkerEntrypoints.ml
+
+    COPY ./scripts/compile-ligo.rb ./scripts/
+    COPY ./scripts/generate-ligo.sh ./scripts/
+    COPY ./patches/e2e-tests-hack.patch .
+
+    ARG E2E_TESTS_HACK=""
+
+    # Note: using bash if-then here instead of earthly's IF-END because the earthly
+    # version was flaky as of version v0.5.23
+    RUN bash -c 'if [ "$E2E_TESTS_HACK" = "true" ]; then patch -p1 <e2e-tests-hack.patch; fi'
+
+    RUN ./scripts/generate-ligo.sh
+    RUN ./scripts/compile-ligo.rb
+
+    SAVE ARTIFACT ./generated/ligo /ligo
+    SAVE ARTIFACT ./generated/michelson /michelson
+    SAVE ARTIFACT ./generated/ligo AS LOCAL ./generated/ligo
+    SAVE ARTIFACT ./generated/michelson AS LOCAL ./generated/michelson
+    SAVE IMAGE --push ghcr.io/tezos-checker/checker/earthly-cache:build-ligo
+
+test-ocaml:
+    FROM +build-ocaml
+    COPY ./scripts/ensure-unique-errors.sh ./scripts/
+    RUN bash ./scripts/ensure-unique-errors.sh
+    RUN opam exec -- dune runtest .
+
+test-ocaml-fast:
+    FROM +build-ocaml
+    COPY ./scripts/ensure-unique-errors.sh ./scripts/
+    RUN bash ./scripts/ensure-unique-errors.sh
+    RUN opam exec -- dune build @run-fast-tests
+
+test-coverage:
+    FROM +build-ocaml
+    COPY ./scripts/ensure-unique-errors.sh ./scripts/
+    RUN bash ./scripts/ensure-unique-errors.sh
+    RUN opam exec -- dune runtest --instrument-with bisect_ppx --force .
+    RUN opam exec -- bisect-ppx-report html
+    RUN echo "$(opam exec -- bisect-ppx-report summary --per-file)"
+    RUN opam exec -- bisect-ppx-report summary --per-file \
+	  | awk '{ match($0, "^ *([0-9.]+) *% *[^ ]* *(.*)$", res); print res[1] "|" res[2] }' \
+	  | jq -R "split(\"|\") | { \
+	      \"value\": .[0] | tonumber, \
+	      \"key\": (.[1] | if . == \"Project coverage\" then \"TOTAL\" else ltrimstr(\"src/\") end) \
+	    }" \
+	  | jq --sort-keys -s 'from_entries' \
+	  | tee test-coverage.json
+    SAVE ARTIFACT _coverage AS LOCAL ./_coverage
```

Would be nice to have the json coverage locally as well:
```suggestion
    SAVE ARTIFACT _coverage AS LOCAL ./_coverage
    SAVE ARTIFACT test-coverage.json AS LOCAL ./test-coverage.json
```